cmd/image/sync/sync_test.go

```diff
@@ -0,0 +1,276 @@
+				mockSyncClient := mocksync.NewMockSyncClient(mockCtrl)
+				return mockSyncClient
+			},
+			instanceList: func(mockSyncClient *mocksync.MockSyncClient) []InstanceItem {
+				return mockCreateInstances(mockSyncClient)
+			},
+			spec: func() []pkg.Spec {
+				return mockCreateSpec()
+			},
+			setup: func(mockSyncClient *mocksync.MockSyncClient) {
+				mockSetBucketLocationConstraint(mockSyncClient, NoOfSources, true, "")
+				mockSetSelectedObjects(mockSyncClient, 0, NoOfSources)
+				mockSetSelectedObjects(mockSyncClient, 0, NoOfSources)
+				mockSetBucketLocationConstraint(mockSyncClient, NoOfSources*NoOfTargetsPerSource, true, "")
+				mockSetCopyObjectToBucket(mockSyncClient, 0, "")
+
+			},
+			expectedError: "",
+		},
+
+		{
+			name: "Bucket Location constraint verification fails for source",
+			mockSyncClient: func(mockCtrl *gomock.Controller) *mocksync.MockSyncClient {
+				mockSyncClient := mocksync.NewMockSyncClient(mockCtrl)
+				return mockSyncClient
+			},
+			instanceList: func(mockSyncClient *mocksync.MockSyncClient) []InstanceItem {
+				return mockCreateInstances(mockSyncClient)
+			},
+			spec: func() []pkg.Spec {
+				return mockCreateSpec()
+			},
+			setup: func(mockSyncClient *mocksync.MockSyncClient) {
+				mockSetBucketLocationConstraint(mockSyncClient, 1, false, "Failed to verify bucket location constraint")
+			},
+			expectedError: "Failed to verify bucket location constraint",
+		},
+
+		{
+			name: "Bucket Location constraint verification fails for a target",
+			mockSyncClient: func(mockCtrl *gomock.Controller) *mocksync.MockSyncClient {
+				mockSyncClient := mocksync.NewMockSyncClient(mockCtrl)
+				return mockSyncClient
+			},
+			instanceList: func(mockSyncClient *mocksync.MockSyncClient) []InstanceItem {
+				return mockCreateInstances(mockSyncClient)
+			},
+			spec: func() []pkg.Spec {
+				return mockCreateSpec()
+			},
+			setup: func(mockSyncClient *mocksync.MockSyncClient) {
+				mockSetBucketLocationConstraint(mockSyncClient, NoOfSources, true, "")
+				mockSetSelectedObjects(mockSyncClient, NoOfObjects, NoOfSources)
+				mockSetSelectedObjects(mockSyncClient, NoOfObjects, 1)
+				mockSetBucketLocationConstraint(mockSyncClient, 1, false, "Failed to verify bucket location constriant")
+			},
+			expectedError: "Failed to verify bucket location constriant",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// creating mock controller object
+			mockCtrl := gomock.NewController(t)
+			defer mockCtrl.Finish()
+
+			// Create object for mock client
+			mockSyncClient := test.mockSyncClient(mockCtrl)
+
+			// test case setup
+			test.setup(mockSyncClient)
+
+			// generating spec slice
+			spec := test.spec()
+
+			// generating necessary instance slice
+			instanceList := test.instanceList(mockSyncClient)
+
+			// test case verification section
+			err := syncObjects(spec, instanceList)
+			if test.expectedError != "" {
+				assert.Error(t, err, test.expectedError)
```

Comment:
assert.Error() won't compare the error, you may need to use `ErrorContains` for error checking in err

Reply:
ErrorContains not available in testify/assert method. Same was implemented using require.
```
require.Error(t, err)
require.Contains(t, err.Error(), test.expectedError)
```

Reply:
how about this one:

https://github.com/stretchr/testify/blob/ab6dc3262822ed562480c19876b0257ace761e3e/assert/assertions.go#L1391

Reply:
I have installed testify version : v1.7.0
When I run with the below code
```
assert.ErrorContains(t, err, test.expectedError)
```

I am getting undefined error:
```
# github.com/ppc64le-cloud/pvsadm/cmd/image/sync [github.com/ppc64le-cloud/pvsadm/cmd/image/sync.test]
./sync_test.go:213:5: undefined: assert.ErrorContains
FAIL	github.com/ppc64le-cloud/pvsadm/cmd/image/sync [build failed]

```

But, I can see the method in GitHub.

Reply:
Seems like it is not yet released in the latest release!

Reply:
fixed the issue in the code with require package mentioned as above.

Reply:
Resolving as discussed.